File: etcd-recipes-examples/src/main/java/io/etcd/recipes/examples/basics/SetValueWithTxn.java
/*
 * Copyright © 2020 Paul Ambrose (pambrose@mac.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.etcd.recipes.examples.basics;

import com.google.common.collect.Lists;
import io.etcd.jetcd.Client;

import java.util.List;

import static io.etcd.recipes.common.ClientUtils.connectToEtcd;
import static io.etcd.recipes.common.KVUtils.*;
import static io.etcd.recipes.common.TxnUtils.*;
import static java.lang.String.format;

public class SetValueWithTxn {
  private static final List<String> urls = Lists.newArrayList("http://localhost:2379");
  private static final String path = "/txnexample";
  private static final String keyval = "foobar";

  public static void main(String[] args) {
    try (Client client = connectToEtcd(urls)) {
      System.out.println("Deleting keys");
      deleteKeys(client, path, keyval);

      System.out.println(format("Key present: %s", isKeyPresent(client, keyval)));
      checkForKey(client);
      System.out.println(format("Key present: %s", isKeyPresent(client, keyval)));
      putValue(client, path, "Something");
      checkForKey(client);
    }
  }

  private static void checkForKey(Client client) {
    transaction(client, (txn) -> {
      txn.If(getDoesExist(path));
      txn.Then(setTo(keyval, format("Key %s found", path)));
      txn.Else(setTo(keyval, format("Key %s not found", path)));
      return txn;
    });

    System.out.println(format("Debug value: %s", getValue(client, keyval, "not_used")));
  }
}

File: etcd-recipes-examples/src/main/kotlin/io/etcd/recipes/examples/basics/SetValueWithTxn.kt
/*
 * Copyright © 2026 Paul Ambrose
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@file:Suppress("UndocumentedPublicClass", "UndocumentedPublicFunction")

package io.etcd.recipes.examples.basics

import io.etcd.jetcd.Client
import io.etcd.recipes.common.connectToEtcd
import io.etcd.recipes.common.deleteKeys
import io.etcd.recipes.common.doesExist
import io.etcd.recipes.common.getValue
import io.etcd.recipes.common.isKeyPresent
import io.etcd.recipes.common.putValue
import io.etcd.recipes.common.setTo
import io.etcd.recipes.common.transaction
import io.github.oshai.kotlinlogging.KotlinLogging

fun main() {
  val logger = KotlinLogging.logger {}
  val urls = ["http://localhost:2379"]
  val path = "/txnexample"
  val keyval = "debug"

  fun checkForKey(client: Client) {
    client.transaction {
      If(path.doesExist)
      Then(keyval setTo "Key $path found")
      Else(keyval setTo "Key $path not found")
    }

    logger.info { "Debug value: ${client.getValue(keyval, "not_used")}" }
  }

  connectToEtcd(urls) { client ->
    logger.info { "Deleting keys" }
    client.deleteKeys(path, keyval)

    logger.info { "Key present: ${client.isKeyPresent(keyval)}" }
    checkForKey(client)
    logger.info { "Key present: ${client.isKeyPresent(keyval)}" }
    client.putValue(path, "Something")
    checkForKey(client)
  }
}
